### PR TITLE
[StructuralMechanicsAppliction] Add Isotropic Damage 3D CL

### PR DESCRIPTION
**Pull request by marandra**
This PR adds an Isotropic Damage Hardening/Softening 3D CL to Structural Mechanics Application, and corresponding tests. In a later PR I will submit a 2D version of it.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py [hunk 1]
```diff
@@ -625,6 +633,46 @@ def initialize_reference_stress(self, strain_size):
...
         return self.reference_stress[i]
 
+class DeformationLinearIsotropicDamage(Deformation):
+    def __init__(self):
+        Deformation.__init__(self)
+        self.nr_timesteps = 10
+
+    def set_deformation(self, cl_params, i):
+        self.strain = (i+1)/ self.nr_timesteps * self.initial_strain
+        cl_params.SetStrainVector(self.strain)
```
Comment (PhilippeBussetta) [inline applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:643]: Same that into `DeformationLinearJ2Plasticity`

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py [hunk 2]
```diff
@@ -625,6 +633,46 @@ def initialize_reference_stress(self, strain_size):
...
+
+class DeformationLinearIsotropicDamage3D(DeformationLinearIsotropicDamage):
```
Comment (PhilippeBussetta) [inline applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:645]: Maybe this class could be derived from `DeformationLinearJ2Plasticity`

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+protected:
+
+    ///@name Protected static Member Variables
+    ///@{
+    ///@}
+
+    ///@name Protected member Variables
+    ///@{
+    bool mInelasticFlag; /// Flags when in inelastic regime
+    double r; /// Damage threshold
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h]: can you please you a more descriptive name?
also it should be preceeded by `m`

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+//************************************************************************************
+
+void LinearIsotropicDamage3D::FinalizeSolutionStep(
+    const Properties& rMaterialProperties,
+    const GeometryType& rElementGeometry,
+    const Vector& rShapeFunctionsValues,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    // update of damage threshold
+    r_prev = r;
```
Comment (PhilippeBussetta) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Where `r` is defined?

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py [hunk 5]
```diff
@@ -733,5 +781,29 @@ def __init__(self):
...
+        return properties
+
+class LinearIsotropicDamage3D(LinearIsotropicDamage):
+    def __init__(self):
+        LinearIsotropicDamage.__init__(self)
+        self.dim = 3
+
+    @staticmethod
+    def create_constitutive_Law():
+        return StructuralMechanicsApplication.LinearIsotropicDamage3DLaw()
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:806]: I would say that this should be a general method, or you change the name

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+#if !defined(KRATOS_LINEAR_ISOTROPIC_DAMAGE_3D_LAW_H_INCLUDED)
+#define KRATOS_LINEAR_ISOTROPIC_DAMAGE_3D_LAW_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/checks.h"
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h]: This should be imported in the cpp, from what I can see it is not used in the header file

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+
+    if (rMaterialProperties[ISOTROPIC_HARDENING_MODULUS] > 0 &&
+        rMaterialProperties[INFINITY_YIELD_STRESS] <= rMaterialProperties[YIELD_STRESS])
+        KRATOS_ERROR << "If ISOTROPIC_HARDENING_MODULUS is positive, "
+            "INFINITY_YIELD_STRESS must be greater than YIELD_STRESS" << std::endl;
+
+    if (rMaterialProperties[ISOTROPIC_HARDENING_MODULUS] < 0 &&
+        rMaterialProperties[INFINITY_YIELD_STRESS] >= rMaterialProperties[YIELD_STRESS])
+        KRATOS_ERROR << "If ISOTROPIC_HARDENING_MODULUS is negative, "
+                "INFINITY_YIELD_STRESS must be lesser than YIELD_STRESS" << std::endl;
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: smaller

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+{
+    if (rThisVariable == STRAIN_ENERGY){
+        Vector& strain_vector = rParameterValues.GetStrainVector();
+        if (rParameterValues.GetProcessInfo().Has(INITIAL_STRAIN)) {
+            noalias(strain_vector) += rParameterValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
+        const Properties& rMaterialProperties = rParameterValues.GetMaterialProperties();
+        Matrix& constitutive_matrix = rParameterValues.GetConstitutiveMatrix();
+        CalculateConstitutiveMatrix(constitutive_matrix, rMaterialProperties);
+        const double q = CalculateQ(r_prev, rMaterialProperties);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: those variable names are not very descriptive

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+            noalias(strain_vector) += rValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
+        Matrix& constitutive_matrix = rValues.GetConstitutiveMatrix();
+
+        CalculateConstitutiveMatrix(constitutive_matrix, rMaterialProperties);
+        noalias(stress_vector) = prod(constitutive_matrix, strain_vector);
+        // For use in TRACTION_ONLY case.
+        Vector stress_vector_pos = prod(constitutive_matrix, strain_vector);
+
+        // Uncomment when implemented TRACTION_ONLY
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: comments should be removed, you could restore them later if needed for further implementation

**Review by philbucher — COMMENTED** [review]
Some minor comments but otherwise nice work, no objections from my side

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+        if (rValues.GetProcessInfo().Has(INITIAL_STRAIN)) {
+            noalias(strain_vector) += rValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
+        Matrix& constitutive_matrix = rValues.GetConstitutiveMatrix();
+
+        CalculateConstitutiveMatrix(constitutive_matrix, rMaterialProperties);
+        noalias(stress_vector) = prod(constitutive_matrix, strain_vector);
+        // For use in TRACTION_ONLY case.
+        Vector stress_vector_pos = prod(constitutive_matrix, strain_vector);
+
```
Comment (PhilippeBussetta) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: It is elastic behaviour?
Maybe you can used the function of the elastic CL

Comment (PhilippeBussetta) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Do you really need this vector `stress_vector_pos`?

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py [hunk 11]
```diff
@@ -295,6 +295,14 @@ def test_J2_Plasticity_Plane_Strain_2D(self):
...
         self._generic_constitutive_law_test(model_part, deformation_test)
 
+    def test_Isotropic_Damage_3D(self):
+        # Define a model
+        model_part = KratosMultiphysics.ModelPart("test")
+
+        deformation_test = DeformationLinearIsotropicDamage3D()
+
+        self._generic_constitutive_law_test(model_part, deformation_test)
+
```
Comment (PhilippeBussetta) [inline applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:305]: Maybe you can add more tests (e.g. Uniaxial, shear)

**Review by PhilippeBussetta — COMMENTED** [review]
Nice work, just minor comments.
Maybe we have to think the link with the elastic CLs

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h (hunk 3 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h]: True. Done.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h (hunk 6 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.h]: done

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+    : ConstitutiveLaw(rOther)
+{
+}
+
+//********************************CLONE***********************************************
+//************************************************************************************
+
+ConstitutiveLaw::Pointer LinearIsotropicDamage3D::Clone() const
+{
+    LinearIsotropicDamage3D::Pointer p_clone(new LinearIsotropicDamage3D(*this));
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Better to use make_shared than new, @RiccardoRossi can you confirm?

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 7 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: lesser doesn't sound good to me either, but I think that lesser and greater are for comparing a relative values, and smaller and bigger for absolute ones.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 4 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: is a member defined in the .h file. I have renamed it to a more descriptive name though.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 10 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: stress_vector_pos is not needed now, will be used in the future. Removing it.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 9 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Right. Done.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 8 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Idem

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+    const double H = rMaterialProperties[ISOTROPIC_HARDENING_MODULUS];
+    const double r0 = yield_stress / std::sqrt(young_modulus);
+    const double q_inf = inf_yield_stress / std::sqrt(young_modulus);
+    double q;
+
+    if (r < r0)
+        return r;
+    q = r0 + H * (r - r0);
+    if ((H > 0 && q > q_inf) || (H < 0 && q < q_inf))
+        q = q_inf;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: Be careful with the zero comparison for doubles, maybe better use epsilons

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 7 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: I think is correct

**Review by loumalouomega — DISMISSED** [review]
Look our comments

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 12 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: I confirm.

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 13 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: done

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py (hunk 5 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:806]: The CL tests are based on the recent code reorganization by @PhilippeBussetta . I'd like to discuss it on a separate issue.

File: applications/StructuralMechanicsApplication/tests/test_constitutive_law.py (hunk 2 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/tests/test_constitutive_law.py:645]: As a general comment for the CL tests, I'd prefer more independence between them. Maybe we can discuss about it in later

File: applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp (hunk 12 above)
Comment (marandra) [reply applications/StructuralMechanicsApplication/custom_constitutive/linear_isotropic_damage_3D_law.cpp]: done

**Comment by marandra** [comment]
Thanks for your comments. All of them should be addressed, except for changes to the test_constitutive_law.py, which I'd like to discuss separately.

**Review by PhilippeBussetta — APPROVED** [review]

**Pull request merged**